[Riako] 스프링 카페 1단계 - 회원 가입 및 목록 기능

naneun opened the pull request:
## PR 인사
안녕하세요! Riako 입니다. 🎉
스프링 카페 1단계 - 회원 가입 및 목록 기능 구현하여 PR 드립니다!
구현보다는 학습한 내용들을 최대한 적용해보고자 노력했습니다.
컨트롤러 단위 테스트의 경우 요청과 응답을 검증하는 부분에서 많이 고민하며 작성했습니다.
귀한 시간 내주시어 리뷰해주셔서 정말 감사합니다. 😀

## PR 목적
  >- 스프링 카페 1단계 - 회원 가입 및 목록 기능 완료

## 특이사항
  >- 기능 및 프로그래밍 요구사항 정리는 README.md 에 작성해놓았습니다.
  >- Controller 와 Service 의 경우 Unit 테스트와 통합? 테스트 모두 작성해보았습니다.

## 고민사항
  >- mustache 의 경우 html 에서 {{@index}} 가 적용시켜보고자 했는데 자료도 많이 없고 적용이 잘되지 않았습니다.
  >- 백엔드에서 화면에 출력되는 index 도 고려해야하는 지 여쭤보고 싶습니다!
  >- 단위 테스트의 경우 (예를 들어 UserController 의 단위 테스트) 시나리오를 작성하는 방법에 대해서 조언을 부탁드리고 싶습니다.
     - 예를 들어 회원이 존재하지 않는 상태에서 테스트 코드의 실행 순서를 정하여 회원들을 추가해가며, 실행 당시의 상황에 맞춰 테스트를 하는 것이 올바른지, 아니면 초기에 어느 정도의 환경을 설정해두고 그 상황을 BeforeEach 로 재설정해가며 같은 상황에서 모든 테스트를 검증해야할 지에 관해서 고민했었습니다. 일단 자료를 찾아보기보단 혼자 최대한 생각을 해보고 작성해보았습니다. 해당 부분에 대해서 조언 부탁드리겠습니다 😄

naneun committed:
🎉 프로젝트 생성 및 설정

naneun committed:
✨ 자바 설정 파일 생성 및 User 도메인 생성

naneun committed:
✨ UserRepositry 인터페이스 및 구현체 생성

- UserRepository 인터페이스: User 리스트를 관리하기 위한 인터페이스
- VolatilityUserRepository: User 리스트를 관리를 위해 ArrayList 를 사용하는 일급객체

naneun committed:
✨ UserService 인터페이스 및 구현체 생성

- UserService 인터페이스: User 도메인 관련 비지니스 로직을 담당하는 인터페이스
- VolatilityUserService: UserService 구현체. 현재 UserRepositry 구현체는 하나만 존재하므로 VolatilityUserRepoistory 를 주입받아 사용한다.

naneun committed:
✅ VolatilityUserService, VolatilityUserRepository 테스트 코드 추가

naneun committed:
✨ UserController 클래스 추가

- /user/create 회원가입
- /users 회원목록
- /users/{userId} 회원프로필

naneun committed:
✨ UserExceptionAdvice 클래스 생성

- UserController 에서 발생하는 User 도메인 관련 예외처리를 담당하는 클래스
- 예외가 발생하면 클라이언트에 BadRequest 로 응답한다.

naneun committed:
✅ UserController 테스트 코드 추가

naneun committed:
✨ LogInterceptor 클래스 생성

- 요청시 UserController 에서 해당 요청 정보를 로그로 남기기 위해 인터셉터를 추가

naneun committed:
🎨 UserController URI 개선

naneun committed:
✅ CafeApplication 테스트 코드 추가

naneun committed:
🎨 no newline at end of file 해결

naneun committed:
🎨 전반적인 테스트 코드 개선

naneun committed:
🎨 Controller 테스트에서 static block -> @BeforeAll 메서드로 변경

naneun committed:
✅ UserController getUserProfileFail 메서드 테스트 추가

- 등록되지 않은 회원의 프로필을 요청 시 응답코드 값 검증 테스트

naneun committed:
🎨 테스트 코드 when -> given 으로 변경

wheejuni commented on src/main/java/com/kakao/cafe/advice/UserExceptionAdvice.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice(assignableTypes = UserController.class)
+public class UserExceptionAdvice {
```
헐~ 단계별 요구사항에 충실한 구현을 해주세요 ㅜㅜ

naneun replied:
ㅎㅎ 넵!..

wheejuni commented on src/main/java/com/kakao/cafe/config/MvcConfig.java:
```diff
@@ -0,0 +1,42 @@
+    public void addViewControllers(ViewControllerRegistry registry) {
+        registry.setOrder(Ordered.HIGHEST_PRECEDENCE);
+
+        registry.addViewController("/").setViewName("index");
+
+        registry.addViewController("/users/form").setViewName("user/form");
+        registry.addViewController("/users/list").setViewName("user/list");
+        registry.addViewController("/users/login").setViewName("user/login");
+        registry.addViewController("/users/login_failed").setViewName("user/login_failed");
+        registry.addViewController("/users/profile").setViewName("user/profile");
+
+        registry.addViewController("/questions/form").setViewName("qna/form");
+        registry.addViewController("/questions/show").setViewName("qna/show");
```
개별 컨트롤러에서 `ModelAndView` 타입 리턴으로 매핑되지 않던가요?

naneun replied:
루카스에 올라온 예시 코드 패턴을 따라적었던 거였는데 그대로 둔 거 같습니다!.. 😅

wheejuni commented on src/main/java/com/kakao/cafe/domain/User.java:
```diff
@@ -0,0 +1,103 @@
+            return new User(this);
+        }
+    }
+    /* End */
```
헉 이건 뭐죠?

naneun replied:
빌더 패턴 작성하는 연습을 해봤습니다!.. 😅

wheejuni commented on src/main/java/com/kakao/cafe/interceptor/LogInterceptor.java:
```diff
@@ -0,0 +1,49 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class LogInterceptor implements HandlerInterceptor {
```
이건 좀 ㅜㅜ 뭘 하고싶은지는 알겠는데, 너무 과하다는 생각이 듭니다.
지금 단계에선 스프링에 너무 침습적인 학습을 안 했으면 좋겠고요, 이런 연습도 지금 단계에서 너무 과해요 ㅜㅜ

밑에 `MethodParameter` 쓰시면서 파라메터까지 로깅하셨는데 굳이 모든 파라메터를 로깅해야 하는 필요가 있는지도 잘 모르겠기는 합니다. 로깅이 필요하다면 각 컨트롤러에서의 로직을 통해 진행해 주시면 좋겠습니다.

naneun replied:
넵! 몽땅 수정해놓도록 하겠습니다!

wheejuni commented on src/main/java/com/kakao/cafe/repository/UserRepository.java:
```diff
@@ -0,0 +1,13 @@
+public interface UserRepository {
+
+    List<User> selectAll();
+    void insertUser(User user);
```
JPA repository의 `.save()` 가 어떤 리턴타입을 갖는지도 찾아보세요~

naneun replied:
JPA 공부해보도록 하겠습니다! 감사합니다!

wheejuni reviewed (changes requested):
수고 많으셨습니다. 💯 
다음부터는 단계별 요구사항에 맞는 진행을 부탁드립니다~

wheejuni commented:
백엔드에서 화면에 출력되는 인덱스를 고려해야한다기보단 백엔드는 백엔드에서 가져다줄 수 있는 데이터를 충분히 구조화해서 내려주고, FE측에서 그런 요구사항이 있다면 기술적 제약 조건이 있는게 아니라면 그거에 맞춰 대응을 해줘야겠죠? ㅎㅎ

컨트롤러 테스트 시의 데이터 초기화 및 적재는 사실 저도 할때마다 좀 짜증스럽기도 하고 귀찮기도 한 부분인데요; 이건 영상 자료로 대체합니다. https://www.youtube.com/watch?v=ITVpmjM4mUE 를 참고해주세요~

naneun commented:
> 백엔드에서 화면에 출력되는 인덱스를 고려해야한다기보단 백엔드는 백엔드에서 가져다줄 수 있는 데이터를 충분히 구조화해서 내려주고, FE측에서 그런 요구사항이 있다면 기술적 제약 조건이 있는게 아니라면 그거에 맞춰 대응을 해줘야겠죠? ㅎㅎ

- 넵! 새겨 듣겠습니다! 😄

> 
> 컨트롤러 테스트 시의 데이터 초기화 및 적재는 사실 저도 할때마다 좀 짜증스럽기도 하고 귀찮기도 한 부분인데요; 이건 영상 자료로 대체합니다. https://www.youtube.com/watch?v=ITVpmjM4mUE 를 참고해주세요~

- 오오 영상 감사드립니다! 감사히 보겠습니다! 🤩

naneun committed:
🚑 LogInterceptor 삭제

- UserController 에서 Logger 객체를 사용하여 로그를 남기도록 변경
- HttpServletRequest, HttpServletResponse 객체를 사용하기 위해 파라미터 추가

naneun committed:
🚑 UserController 메서드 파라미터 순서 변경

naneun committed:
🚑 ExceptionAdvice 삭제

- Controller 내에서 ExceptionHandler 를 설정하여 사용하도록 변경

naneun committed:
🚑 UserRepository insertUser 메서드 반환 값 변경

- JPA save 메서드의 구현체를 참고하여 반환 값을 void -> Optional<User> 변경

naneun committed:
✅ insertUser 시그니처 변경에 따른 테스트 코드 수정

naneun committed:
✨ list.html 에서 회원목록 출력시 함께 표시될 index 추가

naneun committed:
🎨 Repository 의 컬렉션 타입을 ArrayList -> Vector 로 변경

- 동기화를 위해 Vector Collection 을 사용

naneun committed:
🎨 User(Builder builder) 생성자 위치 변경

- 클래스 구현 순서 규칙 적용

naneun committed:
🎨 users.size(): ArrayList 의 크기로 아이디를 할당하도록 변경

naneun committed:
🚑 insertUser 메서드 동시성 레벨 변경

- 객체 레벨 동시성에서 메서드 레벨 동시성으로 변경

naneun committed:
🎨 UserRepository 인스턴스 메서드 이름 변경

- insertUser -> save, selectAll, selectUser -> findAll, findOne

naneun committed:
✨ 3단계 미션을 위한 h2 database 연동

naneun committed:
🎨 UserController @ExceptionHandler 접근제어자 변경

naneun committed:
🎨 spring.datasource.url 값 변경

naneun committed:
🐛 index.html 에서 사용하는 url 재정리

naneun committed:
✨ updateForm.html 생성

naneun committed:
🎨 index.html 에서 사용되는 url 정리 마무리

naneun committed:
♻️ Repository 인터페이스 추가

- JPA save 메서드를 참고하여 Repository 구조 변경
- 추상화 클래스를 추가하여 데이터 저장소를 구체화(1단계)
- 추상화 클래스를 구체화한 클래스를 사용하여 동작 구현(2단계)

naneun committed:
✅ 전반적인 코드 구조 변경에 따른 테스트 코드 수정

- ObjectMapper convertValue 메서드 사용시 인자로 new TypeReference<>() 전달하기
- UserService 의 메서드 명을 전반적으로 변경

naneun committed:
♻️ ENTITY_STATUS enum 클래스로 인한 save 메서드 구조 변경

- Repository 인터페이스에 ENTITY_STATUS enum 클래스를 추가하여 엔티티 상태값을 설정할 수 있도록 변경

naneun committed:
♻ Repository 인터페이스의 제네릭화로 인한 비용 처리

- VolatilityUserService 의 의존성 주입 대상으로 Repository 로 변경
- 자바의 타입 삭제 방식으로 인한 객체의 런타임 타입 다운케스팅

naneun committed:
♻  타입 매개 변수를 명시하여 Raw 타입 제거

naneun committed:
✨ 회원정보 수정 기능 추가 (비밀번호 확인 제외)

naneun committed:
✨ 회원정보 수정 기능 완성 (비밀번호 확인 추가)
